Documentation/backend/typescript/mongodb.md

```diff
@@ -0,0 +1,140 @@
+This is then used by custom middlewares to resolve resources correctly.
+
+## MongoDbReadModelsConfiguration
+
+The actual configuration from a `resources.json` file is read and parsed and translated into types that can be leveraged.
+ReadModel configurations for MongoDb is defined in a type called `MongoDbReadModelsConfiguration` that looks like the following:
+
+```typescript
+{
+    host: string;
+    database: string;
+    useSSL: boolean;
+}
+```
+
+## Express Middleware
+
+For the serialization subsystem to work properly out of the box, it relies on being aware of the type mapped to a collection
+for the web request it is being used in. To achieve this there is an Express Middleware that can be used.
+The default setup gets set up with this, but if you have your own setup and want to leverage
+this functionality - you can do so through the following:
+
+```typescript
+import express from 'express';
+import {MongoDbMiddleware } from '@dolittle/vanir-backend/dist/mongodb';
+
+const app = express();
+app.use(MongoDbMiddleware);
+```
+
+## Working with collections
+
+When your system needs to access a specific collection, it can go through the `IMongoDatabase` interface to do so.
+The advantage of doing so is that it will hook up the serialization sub system to work unobtrusively for you as a developer.
+
+That way you don't have to hand-code conversions going back and forth to MongoDB, but rely on serializers already having
+been set up to do the job for you.
+
+The API offers a way to get collections by associating a schema/document type. This is the type it leverages to make
+the serialization sub system work.
+
+Lets say you have an API controller:
+
+```typescript
+import { IMongoDatabase } from '@dolittle/vanir-backend';
+import { guid } from '@dolittle/vanir-backend/dist/mongodb';
+import { Controller, Get, Route } from '@tsoa/runtime';
+import { injectable } from 'tsyringe';
+import { Guid } from '@dolittle/rudiments';
+
+class MyType {
+
+    @guid()
+    _id!: Guid;
+
+    something: number = 0;
+}
+
+@Route('/api/typescript/my-controller')
+@injectable()
+export class MyController extends Controller {
+
+    constructor(private readonly _db: IMongoDatabase) {
+        super();
+    }
+
+    @Post()
+    async insertThings(): Promise<void> {
+        const collection = await this._db.collection(MyType);
+        await collection.insertOne({
+            _id: Guid.create(),
+            something: 42
+        });
+    }
+
+    @Get()
+    async doStuff(): Promise<MyType[]> {
+        const result = await collection.find().toArray();
+        return result;
+    }
+}
+```
+
+## Custom types
+
+The serialization subsystem introduces the concept of a `CustomType` - these are the ones used to convert to and from
+MongoDBs BSON binary.
+
+Creating your own custom type is easy, all you need to do is to implement the abstract class `CustomType`:
```

Comment:
Great question - I think the documentation could be clearer on this - I’ll add more clarity around it.
CustomTypes are only for those occassions where you want it stored in a particular way rather than what the default behavior is. If you take the Guid that we have a custom type for, if we didn’t have the custom type - it would by default recursively serialize the class Guid and end up with all the bytes as a byte array plus any other properties on that type. While what we’re looking for for that particular type is a way to be able to query it easily in Mongo and also easily read it (or more easily - Guids are hard to read anyways :)).

A good example of a CustomType we probably would like to have is for the DateTime type we’ve been using. Serializing it by default would end up with a structure like this being saved:

```json
{
    «year»: 2021,
    «month»: 2,
    «day»: 1,
    «hour»: 13,
    «minute»: 37,
    «seconds»: 37,
    «milliseconds»: 654
}
```

While what would be nice is that it ended up like the date type in MongoDB. 

I’ve added an issue for the DateTime type : #119 

I think for the most part CustomTypes aren’t needed.

Hope that made sense.